### PR TITLE
vine: handle failed library tasks

### PR DESCRIPTION
**Pull request by JinZhou5042**
## Proposed Changes

There are two cases that a library process may fail:
1. the library process exits with exit code 1 due to bugs in the source code
2. the library process no longer responses due to some environment issues

In the current implementation, handle_completed_tasks will detect if a Unix process has finished with a specific exit status. If it's a library process, it calls handle_failed_library_process for cleanup.

For the second reason, the library process may become a zombie and stay in the OS, preventing waiting function calls from executing. To check if a library process is active, I opened another pipe called heartbeat_pipe between the worker and the library. The library will send a heartbeat message to the worker every 5 seconds, so the worker knows it's active. If the worker doesn't receive any messages for over 30 seconds, it will call handle_failed_library_process to manage the library failure.

In terms of handling the library failures, the worker will 
- Forsake the tasks that are running on it and in the waiting queue that can't find a suitable library
- Mark it as a failed status, move it to the `procs_complete` table and report to the manager

Upon receiving the failure of a library, the manager will:
- Set its status to `VINE_TASK_RETRIEVED`
- Increment the `library_failed_count` for the library template
- Wait for a moment (15s) and duplicate another library from the template
- If a library (template) fails over 15 times, then remove this library permanent 
- Assist users in debugging by printing messages to let them enable watch-library-logfiles or redirect to the target logfile.

Specifically, if the users tune the `watch-library-logfiles` parameter in their Python code:
 
```
q.tune('watch-library-logfiles', 1)
```
Then help them find the logfile
Screenshot from 2024-05-31 16-46-03

If not, then help them tune this parameter:

Screenshot from 2024-05-31 16-50-07

Once tuned `watch-library-logfile`, the remote library output files are transferred into `library_logs`, each with the name format `library-[1-x].debug.log`. 

```
staging
vine-logs
library_logs/
├── library-1.debug.log
├── library-2.debug.log
├── library-3.debug.log
└── library-4.debug.log
```

One of them has this looking:

```
06/05/24 12:37:28.003615 library task starts running in process 1181719
06/05/24 12:37:28.003649 hostname             crcfe02.crc.nd.edu
06/05/24 12:37:28.003659 task id              1201
06/05/24 12:37:28.003665 worker pid           1180379
06/05/24 12:37:28.003670 library pid          1181719
06/05/24 12:37:28.003675 input fd             6
06/05/24 12:37:28.003679 output fd            9
06/05/24 12:37:28.003683 library cores        16
06/05/24 12:37:28.003688 function slots       16
06/05/24 12:37:28.003692 thread limit         1
06/05/24 12:37:28.008669 functions installed
def my_func(size=100):
    import numpy
    import pandas
    import sklearn.preprocessing
    import sklearn.decomposition
    import scipy.stats
    import scipy.special
    import matplotlib.pyplot
    import hashlib
    import os
    print('hello')
    matrix_a = numpy.random.rand(int(size / 2), int(size / 2))
    matrix_b = numpy.random.rand(int(size / 2), int(size / 2))
    data = numpy.matmul(matrix_a, matrix_b)
    data = numpy.random.randn(size, 2) * 200 + numpy.random.rand(size, 2) * 100
    df = pandas.DataFrame(data, columns=['A', 'B'])
    df_normalized = df.apply(scipy.stats.zscore)
    scaler_standard = sklearn.preprocessing.StandardScaler().fit_transform(df_normalized)
    scaler_minmax = sklearn.preprocessing.MinMaxScaler().fit_transform(df_normalized)
    pca = sklearn.decomposition.PCA(n_components=2)
    pca_result = pca.fit_transform(scaler_standard)
    transformed_data = scipy.special.expit(pca_result)
    matplotlib.pyplot.figure(figsize=(12, 6))
    matplotlib.pyplot.subplot(1, 1, 1)
    matplotlib.pyplot.scatter(pca_result[:, 0], pca_result[:, 1], c='blue', label='PCA Result')
    matplotlib.pyplot.title('PCA Plot')
    unique_string = str(transformed_data.tolist()) + str(pca_result.tolist()) + str(scaler_minmax.tolist()) + str(df_normalized.values.tolist())
    hash_object = hashlib.sha256(unique_string.encode())
    hash_string = hash_object.hexdigest()
    return hash_string
def estimate_pi(num_samples=30000000):
    import random
    points_inside_circle = 0
    for _ in range(num_samples):
        x, y = (random.uniform(-1, 1), random.uniform(-1, 1))
        distance_to_center = x ** 2 + y ** 2
        if distance_to_center <= 1:
            points_inside_circle += 1
    pi_estimate = 4 * points_inside_circle / num_samples
    return pi_estimate
06/05/24 12:37:28.008731 0 functions running concurrently
06/05/24 12:37:28.020040 TASK 127 my_func arrives, starting to run in process 1181725
06/05/24 12:37:28.021003 TASK 230 my_func arrives, starting to run in process 1181726
06/05/24 12:37:28.021670 TASK 333 my_func arrives, starting to run in process 1181727
06/05/24 12:37:28.022866 TASK 436 my_func arrives, starting to run in process 1181728
06/05/24 12:37:28.024150 TASK 539 my_func arrives, starting to run in process 1181729
06/05/24 12:37:28.025330 TASK 642 my_func arrives, starting to run in process 1181730
06/05/24 12:37:28.025691 TASK 745 my_func arrives, starting to run in process 1181731
06/05/24 12:37:28.027127 TASK 848 my_func arrives, starting to run in process 1181732
06/05/24 12:37:28.028070 TASK 951 my_func arrives, starting to run in process 1181733
06/05/24 12:37:28.029108 TASK 1054 my_func arrives, starting to run in process 1181734
06/05/24 12:37:28.029330 TASK 1157 my_func arrives, starting to run in process 1181735
06/05/24 12:37:28.030751 TASK 60 my_func arrives, starting to run in process 1181736
06/05/24 12:37:28.031604 TASK 164 my_func arrives, starting to run in process 1181737
06/05/24 12:37:28.032159 TASK 268 my_func arrives, starting to run in process 1181738
06/05/24 12:37:28.033644 TASK 372 my_func arrives, starting to run in process 1181739
06/05/24 12:37:28.034279 TASK 476 my_func arrives, starting to run in process 1181740
06/05/24 12:37:33.038472 16 functions running concurrently
06/05/24 12:37:38.043403 16 functions running concurrently
06/05/24 12:37:43.048395 16 functions running concurrently
06/05/24 12:37:48.053398 16 functions running concurrently
06/05/24 12:37:53.058396 16 functions running concurrently
06/05/24 12:37:57.685211 TASK 1157 finished successfully
06/05/24 12:37:57.756192 TASK 230 finished successfully
06/05/24 12:37:59.401855 TASK 60 finished successfully
06/05/24 12:37:59.409209 TASK 476 finished successfully
06/05/24 12:37:59.503587 TASK 268 finished successfully
06/05/24 12:37:59.547091 13 functions running concurrently
06/05/24 12:37:59.984784 TASK 436 finished successfully
06/05/24 12:38:00.345523 TASK 164 finished successfully
06/05/24 12:38:00.542799 TASK 372 finished successfully
06/05/24 12:38:00.579872 TASK 127 finished successfully
06/05/24 12:38:00.834620 TASK 539 finished successfully
06/05/24 12:38:02.878876 TASK 848 finished successfully
06/05/24 12:38:03.738301 TASK 951 finished successfully
06/05/24 12:38:03.849570 TASK 745 finished successfully
06/05/24 12:38:03.918019 TASK 333 finished successfully
06/05/24 12:38:04.396904 TASK 1054 finished successfully
06/05/24 12:38:04.497892 TASK 642 finished successfully
06/05/24 12:38:09.535357 0 functions running concurrently
06/05/24 12:38:14.540542 0 functions running concurrently
06/05/24 12:38:19.545720 0 functions running concurrently
06/05/24 12:38:24.550915 0 functions running concurrently
06/05/24 12:38:29.556120 0 functions running concurrently
06/05/24 12:38:34.561303 0 functions running concurrently
06/05/24 12:38:39.566502 0 functions running concurrently
```

## Merge Checklist

The following items must be completed before PRs can be merge.
Check these off to verify you have completed all steps.

- [ ] `make test`       Run local tests prior to pushing.
- [ ] `make format`     Format source code to comply with lint policies. Note that some lint errors can only be resolved manually (e.g., Python)
- [ ] `make lint`       Run lint on source code prior to pushing.
- [ ] Manual Update     Update the manual to reflect user-visible changes.
- [ ] Type Labels       Select a github label for the type: bugfix, enhancement, etc.
- [ ] Product Labels    Select a github label for the product: TaskVine, Makeflow, etc.
- [ ] PR RTM            Mark your PR as ready to merge.

### CODE REVIEW

File: taskvine/src/worker/vine_process.c [hunk 1]
```diff
@@ -87,8 +87,8 @@ struct vine_process *vine_process_create(struct vine_task *task, vine_process_ty
 	const char *dirtype = vine_process_sandbox_code(p->type);
 
 	p->sandbox = string_format("%s/%s.%d", workspace->workspace_dir, dirtype, p->task->task_id);
-	p->tmpdir = string_format("%s/.taskvine.tmp", p->sandbox);
```
Comment (btovar) [inline taskvine/src/worker/vine_process.c:90]: No need to modify these names, they are written to their own sandbox.

File: taskvine/src/worker/vine_process.c [hunk 2]
```diff
@@ -108,8 +108,8 @@ void vine_process_delete(struct vine_process *p)
 	if (p->task)
 		vine_task_delete(p->task);
 
-	if (p->output_file_name) {
```
Comment (btovar) [inline taskvine/src/worker/vine_process.c:111]: No need to change this attribute name, let's focus on the log files in this pr.

File: taskvine/src/manager/vine_manager.c [hunk 3]
```diff
@@ -4561,6 +4561,14 @@ struct vine_task *send_library_to_worker(struct vine_manager *q, struct vine_wor
 	/* Give it a unique taskid if library fits the worker. */
 	t->task_id = q->next_task_id++;
 
+	/* Watch the output file of every duplicated library instance */
+	char *remote_stdout_filename = string_format(".task%d.stdout", t->task_id);
```
Comment (btovar) [inline taskvine/src/manager/vine_manager.c:4565]: This could simply be `"library.log"`. There is no need to create a specific name remotely.

File: taskvine/src/manager/vine_runtime_dir.c [hunk 4]
```diff
@@ -112,6 +112,12 @@ char *vine_runtime_directory_create()
 	register_staging_dir(tmp);
 	free(tmp);
 
+	tmp = string_format("%s/library-stdout", runtime_dir);
```
Comment (btovar) [inline taskvine/src/manager/vine_runtime_dir.c:115]: Use `vine_get_runtime_library_...(m, NULL)`.

File: taskvine/src/manager/vine_runtime_dir.c [hunk 5]
```diff
@@ -138,6 +144,11 @@ char *vine_get_runtime_path_staging(struct vine_manager *m, const char *path)
 	return string_format("%s/staging%s%s", m->runtime_directory, path ? "/" : "", path ? path : "");
 }
 
+char *vine_get_runtime_path_library_stdout(struct vine_manager *m, const char *path)
```
Comment (btovar) [inline taskvine/src/manager/vine_runtime_dir.c:147]: Change `stdout` to `log`?

File: poncho/src/poncho/library_network_code.py [hunk 6]
```diff
@@ -19,12 +19,26 @@ def library_network_code():
     import cloudpickle
     import select
     import signal
+    import time
+    import builtins
```
Comment (btovar) [inline poncho/src/poncho/library_network_code.py:23]: Instead of `print` to stdout, let's use `sys.write` to `stderr`.

File: poncho/src/poncho/library_network_code.py [hunk 7]
```diff
@@ -19,12 +19,26 @@ def library_network_code():
...
 
     # self-pipe to turn a sigchld signal when a child finishes execution
     # into an I/O event.
     r, w = os.pipe()
 
+    # Ensure each print statement automatically flushes the stdout file,
+    # this is important of tracking the stdout timely
+    original_print = builtins.print
+
+    def flushing_print(*args, **kwargs):
```
Comment (btovar) [inline poncho/src/poncho/library_network_code.py:36]: If using `stderr`, there is no need to activate flushing.

File: poncho/src/poncho/library_network_code.py [hunk 8]
```diff
@@ -142,44 +169,50 @@ def start_function(in_pipe_fd, thread_limit=1):
...
                         # setup stdout/err for a function call so we can capture them.
+                        # redirect the stdout of a specific FunctionCall task into its stdout file
                         stdout_capture_fd = os.open(
                             function_stdout_file, os.O_WRONLY | os.O_CREAT | os.O_TRUNC
                         )
                         os.dup2(stdout_capture_fd, sys.stdout.fileno())
                         os.dup2(stdout_capture_fd, sys.stderr.fileno())
                         os.chdir(function_sandbox)
 
+                        exit_status = 1
```
Comment (btovar) [inline poncho/src/poncho/library_network_code.py:193]: Move this `exit_status` just after the `try`.

File: poncho/src/poncho/library_network_code.py [hunk 9]
```diff
@@ -189,19 +222,18 @@ def start_function(in_pipe_fd, thread_limit=1):
...
 
                         # nothing failed
                         exit_status = 0
+                    except Exception as e:
+                        print_timed_message(f"FUNCTION{function_id} error: execution failed due to {e}")
                     finally:
                         os.close(stdout_capture_fd)
                         os._exit(exit_status)
                 elif p < 0:
-                    print(
```
Comment (btovar) [inline poncho/src/poncho/library_network_code.py:201]: Why is this message removed?

File: poncho/src/poncho/library_network_code.py [hunk 10]
```diff
@@ -298,19 +354,29 @@ def main():
...
             # check if parent exits
             c_ppid = os.getppid()
             if c_ppid != ppid or c_ppid == 1:
+                print_timed_message("library exits with success")
                 exit(0)
 
             # wait for messages from worker or child to return
             rlist, wlist, xlist = select.select([in_pipe_fd, r], [], [], timeout)
 
+            # periodically log the number of concurrent functions
```
Comment (btovar) [inline poncho/src/poncho/library_network_code.py:371]: Not needed. It will create many messages for information that the manager already has.

**Review by btovar — CHANGES_REQUESTED** [review]

**Review by dthain — COMMENTED** [review]
Jin,  think this is going to be a very helpful capability for debugging.  Two suggestions:

1 -  We don't want watching of outputs to be enabled by default, otherwise for large runs, the manager is going to be collecting a huge quantity of unread debug output from every single library.  Instead, just give each library an (unwatched) output file by default.  That will have the nice outcome that failed libraries will return a log file, while successful libraries will not, because they never exit.

Then, you can add a `vine_tune` option `watch-library-logfiles` that will add the `VINE_WATCH` flag in the less common case where you want to see all of the log files as they are written.

2 - Make sure that the log events produced by the library are full of useful information: the pid of processes forked/completed, the name of the function executed, the text of exceptions caught, etc.

**Comment by JinZhou5042** [comment]
> Jin, think this is going to be a very helpful capability for debugging. Two suggestions:
> 
> 1 - We don't want watching of outputs to be enabled by default, otherwise for large runs, the manager is going to be collecting a huge quantity of unread debug output from every single library. Instead, just give each library an (unwatched) output file by default. That will have the nice outcome that failed libraries will return a log file, while successful libraries will not, because they never exit.
> 
> Then, you can add a `vine_tune` option `watch-library-logfiles` that will add the `VINE_WATCH` flag in the less common case where you want to see all of the log files as they are written.
> 
> 2 - Make sure that the log events produced by the library are full of useful information: the pid of processes forked/completed, the name of the function executed, the text of exceptions caught, etc.

Sounds good!

**Comment by JinZhou5042** [comment]
I think my next step would be printing a warning message and directing the users to a specific file in the `while not q.empty()` loop if a remote library instance fails.

**Comment by btovar** [comment]
@JinZhou5042 Since this is going to be an option that API users will have to enable, please ignore my comments about the deletion of some debug messages.

**Comment by JinZhou5042** [comment]
@btovar  Copy that!

**Comment by dthain** [comment]
Jin, if you are having difficulty getting this to pass the build tests then stop by this afternoon and let's look at it together.

**Comment by JinZhou5042** [comment]
@dthain No worries as of now, I am testing some corner cases to make sure that the worker can always detects the exit of a library process!

File: taskvine/src/manager/vine_manager.c [hunk 11]
```diff
@@ -1662,6 +1665,23 @@ static vine_result_code_t get_result(struct vine_manager *q, struct vine_worker_
 	/* Consume the stdout data immediately following this message. */
 	get_stdout(q, w, t, output_length);
 
+	/* If a library task fails, remove this instance  */
+	/* handle failed library instances */
+	if (task_status == VINE_RESULT_LIBRARY_FAILED) {
+		debug(D_VINE, "Task %d library %s failed", t->task_id, t->provides_library);
+		reset_task_to_state(q, t, VINE_TASK_RETRIEVED);
```
Comment (dthain) [inline taskvine/src/manager/vine_manager.c:1672]: Just do the things specific to libraries here, such as counting up the failures and emitting a message.

But don't shortcut the task transitions: we want all tasks to go through the same transitions for all sorts of reasons: accounting, dealing with output files, etc.  Instead of changing the state here, just fall through the code below and do the normal itable_remove and change_task_state.

Comment (dthain) [reply taskvine/src/manager/vine_manager.c:1672]: In fact, don't even bother checking for VINE_RESULT_LIBRARY_FAILED b/c a library that exits "normally" is still failing from our perspective.

File: taskvine/src/manager/vine_manager.c [hunk 12]
```diff
@@ -1662,6 +1665,23 @@ static vine_result_code_t get_result(struct vine_manager *q, struct vine_worker_
...
+		t->refcount--;
+		struct vine_task *original = hash_table_lookup(q->library_templates, t->provides_library);
```
Comment (dthain) [inline taskvine/src/manager/vine_manager.c:1674]: Careful here -- if there are enough failures, the original could be removed.  So you must check for `if(!original)` here and anywhere we do `hash_table_lookup(q->library_templates)`

File: taskvine/src/manager/vine_manager.c [hunk 13]
```diff
@@ -4430,8 +4458,11 @@ const char *vine_result_string(vine_result_t result)
 	case VINE_RESULT_CANCELLED:
 		str = "CANCELLED";
 		break;
-	case VINE_RESULT_LIBRARY_EXIT:
-		str = "LIBRARY_EXIT";
+	case VINE_RESULT_LIBRARY_CANCELLED:
```
Comment (dthain) [inline taskvine/src/manager/vine_manager.c:4461]: Hmm, I don't think that we want to introduce too many specialized states here.
- `LIBRARY_CANCELLED` should be the same thing as `CANCELLED` (since it gets cancelled in the same way and we want to count up statistics in the same way.
- In the same way `LIBRARY_FAILED` is unneeded b/c anytime a library task runs to completion, we should consider it a failure.

File: taskvine/src/manager/vine_manager.c [hunk 14]
```diff
@@ -5425,6 +5494,17 @@ int vine_tune(struct vine_manager *q, const char *name, double value)
 	} else if (!strcmp(name, "option-blocklist-slow-workers-timeout")) {
 		q->option_blocklist_slow_workers_timeout = MAX(0, value); /*todo: confirm 0 or 1*/
 
+	} else if (!strcmp(name, "watch-library-logfiles")) {
+		char *runtime_dir = xxstrdup(getenv("VINE_RUNTIME_INFO_DIR"));
```
Comment (dthain) [inline taskvine/src/manager/vine_manager.c:5498]: Please do the directory creation in `vine_runtime_directory_create` so we can keep track of the expected directories in one place.  (I realize the dir won't get used unless the feature is enabled.)

File: taskvine/src/worker/vine_worker.c [hunk 15]
```diff
@@ -1441,49 +1465,76 @@ static int enforce_worker_promises(struct link *manager)
...
 /*
-Given a freshly started process, wait for it to initialize and send
-back the library startup message with JSON containing the name of
-the library, which should match the task's provides_library label.
+Kill a failed library process.
+This library process may either not existent or no longer responses.
 */
 
-static int check_library_startup(struct vine_process *p)
+static void handle_failed_library_process(struct vine_process *p, struct link *manager)
```
Comment (dthain) [inline taskvine/src/worker/vine_worker.c:1472]: I would like to minimize the number of ways in which libraries are "different" from ordinary tasks.
I think it's better if a library is reaped in exactly the same way as a normal process.

(And the functions left behind will still be forsaken quickly when they discover `process_can_run_eventually`
 is false.

File: taskvine/src/worker/vine_worker.c [hunk 16]
```diff
@@ -1492,27 +1543,47 @@ static void check_libraries_ready()
...
-							"Library %s reports ready to execute functions.",
-							library_process->task->provides_library);
-					library_process->library_ready = 1;
-				} else {
-					/* Kill library if the name reported back doesn't match its name or
-					 * if there's any problem. */
-					debug(D_VINE, "Library verification failed. Killing it.");
-					vine_process_kill(library_process);
-				}
+		/* Check if library has sent its startup message. */
```
Comment (dthain) [inline taskvine/src/worker/vine_worker.c:1552]: Hmm, I am not sure that the heartbeat idea is actually making things better here.

The previous idea was that the library sends a startup message once it has completed its own internal initialization.  (And that could take longer than 30s.)

This is adding more pipes and things to check, but not actually improving upon that detection.

I think you have enough good things in this PR without adding the heartbeat mechanism.

**Review by dthain — CHANGES_REQUESTED** [review]
This is looking pretty good, I have a few small suggestions inline.  But the main thing is that the PR is getting too big, please reverse the changes regarding heartbeats.

File: poncho/src/poncho/library_network_code.py [hunk 17]
```diff
@@ -72,13 +77,21 @@ def remote_wrapper(event):
...
+            all_function_code = []
+
+            for node in ast.walk(tree):
+                if isinstance(node, ast.FunctionDef):
+                    for decorator in node.decorator_list:
+                        if isinstance(decorator, ast.Name) and decorator.id == 'remote_execute':
+                            function_code = ast.get_source_segment(content, node)
+                            all_function_code.append(function_code)
+
+        return "\n".join(all_function_code)
```
Comment (dthain) [inline poncho/src/poncho/library_network_code.py:94]: Does this return the entire source code of the library functions?
I don't think you want to do that -- this could be very large in the general case.
Is your goal here to just clarify what function names are installed?

File: poncho/src/poncho/library_network_code.py [hunk 18]
```diff
@@ -265,48 +321,82 @@ def main():
...
+        stdout_timed_message(f"hostname             {socket.gethostname()}")
+        stdout_timed_message(f"task id              {args.task_id}")
+        stdout_timed_message(f"worker pid           {args.worker_pid}")
+        stdout_timed_message(f"library pid          {os.getpid()}")
+        stdout_timed_message(f"input fd             {args.in_pipe_fd}")
+        stdout_timed_message(f"output fd            {args.out_pipe_fd}")
+        stdout_timed_message(f"library cores        {args.library_cores}")
+        stdout_timed_message(f"function slots       {args.function_slots}")
+        stdout_timed_message(f"thread limit         {thread_limit}")
+        stdout_timed_message(f"functions installed\n{get_installed_function_code()}")
```
Comment (dthain) [inline poncho/src/poncho/library_network_code.py:357]: Again, the names of the functions, or the entire function code?

File: taskvine/src/manager/vine_manager.c [hunk 19]
```diff
@@ -1436,6 +1438,7 @@ static vine_result_code_t get_update(struct vine_manager *q, struct vine_worker_
 
 	int n = sscanf(line, "update %" PRId64 " %s %" PRId64 " %" PRId64, &task_id, path, &offset, &length);
 	if (n != 4) {
+		debug(D_VINE, "WHAT?");
```
Comment (dthain) [inline taskvine/src/manager/vine_manager.c:1441]: Remove

File: taskvine/src/manager/vine_manager.c [hunk 20]
```diff
@@ -5211,6 +5276,9 @@ int vine_cancel_by_task_id(struct vine_manager *q, int task_id)
 		return 0;
 	}
 
+	if (task->type == VINE_TASK_TYPE_LIBRARY_INSTANCE)
```
Comment (dthain) [inline taskvine/src/manager/vine_manager.c:5279]: Don't manipulate the refcount directly.
Just call `vine_task_delete` which removes a refcount and will delete if it goes to zero.
(The idea is that we should have mirror pairs of `vine_task_addref` and `vine_task_delete`

**Review by dthain — CHANGES_REQUESTED** [review]
A few little things here and there to fix.

What's the story with `get_function_code`? Does it really send back the entire function source, or just the names?

File: taskvine/src/manager/vine_manager.c (hunk 19 above)
Comment (JinZhou5042) [reply taskvine/src/manager/vine_manager.c:1441]: Sorry about this...

File: poncho/src/poncho/library_network_code.py (hunk 18 above)
Comment (JinZhou5042) [reply poncho/src/poncho/library_network_code.py:357]: @dthain It logs the entire function code, for example:
```
06/05/24 12:37:28.008669 functions installed
def my_func(size=100):
    import numpy
    import pandas
    import sklearn.preprocessing
    import sklearn.decomposition
    import scipy.stats
    import scipy.special
    import matplotlib.pyplot
    import hashlib
    import os
    print('hello')
    matrix_a = numpy.random.rand(int(size / 2), int(size / 2))
    matrix_b = numpy.random.rand(int(size / 2), int(size / 2))
    data = numpy.matmul(matrix_a, matrix_b)
    data = numpy.random.randn(size, 2) * 200 + numpy.random.rand(size, 2) * 100
    df = pandas.DataFrame(data, columns=['A', 'B'])
    df_normalized = df.apply(scipy.stats.zscore)
    scaler_standard = sklearn.preprocessing.StandardScaler().fit_transform(df_normalized)
    scaler_minmax = sklearn.preprocessing.MinMaxScaler().fit_transform(df_normalized)
    pca = sklearn.decomposition.PCA(n_components=2)
    pca_result = pca.fit_transform(scaler_standard)
    transformed_data = scipy.special.expit(pca_result)
    matplotlib.pyplot.figure(figsize=(12, 6))
    matplotlib.pyplot.subplot(1, 1, 1)
    matplotlib.pyplot.scatter(pca_result[:, 0], pca_result[:, 1], c='blue', label='PCA Result')
    matplotlib.pyplot.title('PCA Plot')
    unique_string = str(transformed_data.tolist()) + str(pca_result.tolist()) + str(scaler_minmax.tolist()) + str(df_normalized.values.tolist())
    hash_object = hashlib.sha256(unique_string.encode())
    hash_string = hash_object.hexdigest()
    return hash_string
def estimate_pi(num_samples=30000000):
    import random
    points_inside_circle = 0
    for _ in range(num_samples):
        x, y = (random.uniform(-1, 1), random.uniform(-1, 1))
        distance_to_center = x ** 2 + y ** 2
        if distance_to_center <= 1:
            points_inside_circle += 1
    pi_estimate = 4 * points_inside_circle / num_samples
    return pi_estimate
```

File: poncho/src/poncho/library_network_code.py (hunk 17 above)
Comment (JinZhou5042) [reply poncho/src/poncho/library_network_code.py:94]: @dthain That's fare, I will replace with only logging the function names.

File: taskvine/src/worker/vine_worker.c [hunk 21]
```diff
@@ -544,7 +544,12 @@ static void reap_process(struct vine_process *p, struct link *manager)
 	gpus_allocated -= p->task->resources_requested->gpus;
 
 	vine_gpus_free(p->task->task_id);
-	vine_sandbox_stageout(p, cache_manager, manager);
+
+	/* For a library process, do not stageout its sandbox until the watched logfile
```
Comment (dthain) [inline taskvine/src/worker/vine_worker.c:548]: Do not add a special case for libraries.  If it has outputs, they should be staged out in the same way as a normal task.  (And he manager already knows how to deal with watched output files at stageout time.)

**Review by dthain — CHANGES_REQUESTED** [review]
Almost done, just a few little things remaining.

Please note that I'm going to merge #3669 shortly, and you will likely have to do some work to rebase this PR after that one.
